io.c

```diff
@@ -4174,33 +4185,101 @@ io_readlines(const struct getline_arg *arg, VALUE io)
 
 /*
  *  call-seq:
- *     ios.each(sep=$/ [, getline_args])          {|line| block } -> ios
- *     ios.each(limit [, getline_args])           {|line| block } -> ios
- *     ios.each(sep, limit [, getline_args])      {|line| block } -> ios
- *     ios.each(...)                             -> an_enumerator
+ *    each_line(sep = $/, **line_opts) {|line| ... }   -> self
+ *    each_line(limit, **line_opts) {|line| ... }      -> self
+ *    each_line(sep, limit, **line_opts) {|line| ... } -> self
+ *    each_line                                   -> enumerator
  *
- *     ios.each_line(sep=$/ [, getline_args])     {|line| block } -> ios
- *     ios.each_line(limit [, getline_args])      {|line| block } -> ios
- *     ios.each_line(sep, limit [, getline_args]) {|line| block } -> ios
- *     ios.each_line(...)                        -> an_enumerator
+ *  Calls the block with each remaining line read from the stream;
+ *  does nothing if already at end-of-file;
+ *  returns +self+.
  *
- *  Executes the block for every line in <em>ios</em>, where lines are
- *  separated by <i>sep</i>. <em>ios</em> must be opened for
- *  reading or an IOError will be raised.
+ *  With no arguments given, reads lines
+ *  as determined by line separator <tt>$/</tt>:
  *
- *  If no block is given, an enumerator is returned instead.
+ *    f = File.new('t.txt')
+ *    f.each_line {|line| p line }
+ *    f.each_line {|line| fail 'Cannot happen' }
  *
- *     f = File.new("testfile")
- *     f.each {|line| puts "#{f.lineno}: #{line}" }
+ *  Output:
  *
- *  <em>produces:</em>
+ *    "First line\n"
+ *    "Second line\n"
+ *    "\n"
+ *    "Fourth line\n"
+ *    "Fifth line\n"
+ *
+ *  With only string argument +sep+ given,
+ *  reads lines as determined by line separator +sep+;
+ *  see {Line Separator}[#class-IO-label-Line+Separator]:
  *
- *     1: This is line one
- *     2: This is line two
- *     3: This is line three
- *     4: And so on...
+ *    f = File.new('t.txt')
+ *    f.each_line('li') {|line| p line }
+ *
+ *  Output:
+ *
+ *    "First li"
+ *    "ne\nSecond li"
+ *    "ne\n\nFourth li"
+ *    "ne\nFifth li"
+ *    "ne\n"
+ *
+ *  The two special values for +sep+ are honored:
+ *
+ *    f = File.new('t.txt')
+ *    # Get all into one string.
+ *    f.each_line(nil) {|line| p line }
+ *
+ *  Output:
+ *
+ *    "First line\nSecond line\n\nFourth line\nFifth line\n"
+ *
+ *    f.rewind
+ *    # Get paragraphs (up to two line separators).
+ *    f.each_line('') {|line| p line }
+ *
+ *  Output:
+ *
+ *    "First line\nSecond line\n\n"
+ *    "Fourth line\nFifth line\n"
+ *
+ *  With only integer argument +limit+ given,
+ *  limits the number of bytes in each line;
+ *  see {Line Limit}}[#class-IO-label-Line+Limit]:
+ *
+ *    f = File.new('t.txt')
+ *    f.each_line(8) {|line| p line }
+ *
+ *  Output:
+ *
+ *    "First li"
+ *    "ne\n"
+ *    "Second l"
+ *    "ine\n"
+ *    "\n"
+ *    "Fourth l"
+ *    "ine\n"
+ *    "Fifth li"
+ *    "ne\n"
+ *
```

Comment:
Likewise, a section about both `sep` and `limit` would be good to add here.

Reply:
Done, plus a couple of others.